File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
private:
    void findParanthesis(int open,int close, vector<string> &result, string temp, int n){
        if(temp.size() == n *2){
            result.push_back(temp);
        }
        if(open<n) findParanthesis(open+1,close,result,temp + "(",n);
        if(close <open) findParanthesis(open,close+1,result,temp+")",n);
    }
public:
    vector<string> generateParenthesis(int n) {
        int open = 0;
        int close = 0;
        vector<string> result;
        string temp = "";
        findParanthesis(open, close, result,temp,n);
        return result;
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {

    private void findParantheses(int open , int close, StringBuffer pair, int n, List<String> result ){
        if(pair.length() == 2*n){
            String temp = pair.toString();
            result.add(temp);
            return;
        }

        if(open < n) {
            pair.append('(');
            findParantheses(open+1,close,pair,n,result);
            pair.deleteCharAt(pair.length()-1);
        }
        if(close < open) {
            pair.append(')');
            findParantheses(open,close+1,pair,n,result);
            pair.deleteCharAt(pair.length()-1);
        }

    }
    public List<String> generateParenthesis(int n) {
        StringBuffer pair = new StringBuffer();
        int open =0;
        int close = 0;
        List<String> result = new ArrayList<>();
        findParantheses(0,0,pair,n,result);
        return result;
    }
}